File: group-service/src/main/java/com/groupservice/service/GroupService.java
package com.groupservice.service;

import java.util.List;
import java.util.UUID;

import org.apache.coyote.BadRequestException;

import com.groupservice.dto.AddMemberRequest;
import com.groupservice.dto.CreateExpenseRequest;
import com.groupservice.dto.CreateGroupRequest;
import com.groupservice.dto.GroupResponse;
import com.groupservice.entity.Group;
import com.groupservice.entity.GroupExpense;
import com.groupservice.dto.GroupMemberResponse;

public interface GroupService {

    Group createGroup(CreateGroupRequest request);

    void addMemberToGroup(UUID groupId, AddMemberRequest request, UUID requestingUserId);

    void createExpense(UUID groupId, CreateExpenseRequest request) throws BadRequestException;

    List<Group> getGroupsByUserId(UUID userId);

    Group getGroupById(UUID groupId);

    List<GroupMemberResponse> getGroupMembers(UUID groupId);

    List<GroupExpense> getGroupExpenses(UUID groupId);

    void deleteGroup(UUID groupId, UUID requestingUserId);

    Group updateGroup(UUID groupId, String name, UUID requestingUserId);

    void removeMember(UUID groupId, UUID userId, UUID requestingUserId);

    void deleteExpense(UUID groupId, UUID expenseId, UUID requestingUserId);

    GroupResponse buildGroupResponse(Group group);

}

File: finsync-frontend/src/services/group-service.ts
import api from "@/lib/api";
import type {
  CreateGroupRequest,
  GroupResponse,
  AddMemberRequest,
  CreateExpenseRequest,
  GroupExpense,
  GroupMember,
} from "@/types";

export const groupService = {
  // ---- Groups ----
  getGroups: async (): Promise<GroupResponse[]> => {
    const response = await api.get<GroupResponse[]>("/api/groups-root");
    return response.data;
  },

  createGroup: async (data: CreateGroupRequest): Promise<GroupResponse> => {
    const response = await api.post<GroupResponse>("/api/groups-root", data);
    return response.data;
  },

  getGroupById: async (id: string): Promise<GroupResponse> => {
    const response = await api.get<GroupResponse>(`/api/groups/${id}`);
    return response.data;
  },

  updateGroup: async (id: string, name: string): Promise<GroupResponse> => {
    const response = await api.put<GroupResponse>(`/api/groups/${id}`, { name });
    return response.data;
  },

  deleteGroup: async (id: string): Promise<void> => {
    await api.delete(`/api/groups/${id}`);
  },

  // ---- Members ----
  getMembers: async (groupId: string): Promise<GroupMember[]> => {
    const response = await api.get<GroupMember[]>(`/api/groups/${groupId}/members`);
    return response.data;
  },

  addMember: async (groupId: string, data: AddMemberRequest): Promise<void> => {
    await api.post(`/api/groups/${groupId}/members`, data);
  },

  removeMember: async (groupId: string, userId: string): Promise<void> => {
    await api.delete(`/api/groups/${groupId}/members/${userId}`);
  },

  // ---- Expenses ----
  getExpenses: async (groupId: string): Promise<GroupExpense[]> => {
    const response = await api.get<GroupExpense[]>(`/api/groups/${groupId}/expenses`);
    return response.data;
  },

  createExpense: async (groupId: string, data: CreateExpenseRequest): Promise<void> => {
    await api.post(`/api/groups/${groupId}/expenses`, data);
  },
};
